wrappers/rust/icicle-cuda-runtime/src/memory.rs

```diff
@@ -244,6 +264,16 @@ impl<T, const D_ID: usize> DeviceVec<T, D_ID> {
             )))))
         }
     }
+
+    pub fn cuda_malloc_for_device(count: usize, device_id: usize) -> CudaResult<Self> {
+        check_device(device_id);
+        Self::cuda_malloc(count)
+    }
+
+    pub fn cuda_malloc_async_for_device(count: usize, stream: &CudaStream, device_id: usize) -> CudaResult<Self> {
+        check_device(device_id);
+        Self::cuda_malloc_async(count, stream)
+    }
```

Comment:
Whats the use case here for passing a specific deviceId if we require it to match the current device's id?

Reply:
This version is preferred by @vhnatyk over implicit choice of device id in malloc.

Reply:
tl;dr - not that it's preferred but seems aligned with everything else. Longer version: the check ensures we maintain correct device_id through our calls - since we don't have implicit `device_id` management elsewhere except here (by my legacy wip implementation and missed on review?). Maybe we can think of fully implicit or umm _automated_ `device_id `management inherent from the call to  `get_device()` on the current thread - but since we have the field in `DeviceContext` - that makes it redundant and anyway feels like can cause multiple bugs? I guess I had to provide basic description in the doc - but I somehow hoped it will emerge from everyone's edit, since it was so intensively discussed:) have to update doc with the current version of API and diagrams

Reply:
I think internal management of `device_id` by thread makes the most sense. `DeviceContext` can also get set this way by using `get_device()`. It might be more bug prone on our end but I think it will be less error prone on the user end which we don't have control over